engine/include/cubos/engine/gizmos/gizmos.hpp

```diff
@@ -0,0 +1,114 @@
+
+    /// @brief Resource which queues for drawing gizmos, basic primitives useful for debugging and tools.
+    ///
+    /// @ingroup gizmos-plugin
+    class Gizmos final
+    {
+    public:
+        /// @brief Sets the color to be used when drawing any subsequent gizmos.
+        /// @param color Color to be used.
+        void color(const glm::vec3& color);
+
+        /// @brief Draws a line gizmo.
+        /// @param id Identifier of the gizmo.
+        /// @param from One of the ends of the line to be drawn.
+        /// @param to The other end of the line to be drawn.
+        /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
+        /// frame.
+        void drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float lifespan = 0);
+
+        /// @brief Draws a filled box gizmo.
+        /// @param id Identifier of the gizmo.
+        /// @param corner One of the corners of the box to be drawn.
+        /// @param oppositeCorner The opposite corner of the box to be drawn.
+        /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
+        /// frame.
+        void drawBox(const std::string& id, glm::vec3 corner, glm::vec3 oppositeCorner, float lifespan = 0);
+
+        /// @brief Draws a wireframe box gizmo.
+        /// @param id Identifier of the gizmo.
+        /// @param corner One of the corners of the box to be drawn.
+        /// @param oppositeCorner The opposite corner of the box to be drawn.
+        /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
+        /// frame.
+        void drawWireBox(const std::string& id, glm::vec3 corner, glm::vec3 oppositeCorner, float lifespan = 0);
+
+        /// @brief Class that describes a type of gizmo
+        class GizmoBase
+        {
+        protected:
+            const std::string& mId; ///< Gizmo identifier
+            glm::vec3 mColor;       ///< Color of the gizmo
+            float mLifespan;        ///< Time in seconds the gizmo has left to live
+
+        public:
+            GizmoBase(const std::string& id, glm::vec3 color, float lifespan)
+                : mId(id)
+                , mColor(color)
+                , mLifespan(lifespan)
+            {
+            }
+
+            /// @brief draws the gizmo to screen
+            /// @param renderer the GizmosRenderer in use
+            virtual void draw(GizmosRenderer renderer) = 0;
+
+            /// @brief decreases the time the gizmo has left before it is destroyed
+            /// @param delta the time in seconds since the last frame
+            bool decreaseLifespan(float delta)
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
+
+            virtual ~GizmoBase() = default;
+        };
+
+        std::vector<std::shared_ptr<GizmoBase>> gizmosVector; ///< Queued gizmos to be drawn.
```

Comment:
Why not define `GizmoBase` in a source file?

Reply:
Ouh, usually that wouldn't be necessary since pointers support forward declarations. Ig the STL doesn't support it :pensive:. If thats the case, at least make it private, and make this a friend of GizmosRenderer using `friend class GizmosRenderer;`

Reply:
I would have to move the iteration code from a system to the GizmosRenderer class, is that preferable?

Reply:
Whoops. Yeah lets leave it as is then.

Reply:
Could you rename `GizmoBase` to `Gizmo` though? From outside we don't see the implementations anyway

Reply:
Alrighty. Was trying to avoid confusion between Gizmo and Gizmos, but you're right, it's private